### PR TITLE
#498: resolve variables on load #691: solve sub-node merging

### LINKED ISSUE #498: Improvement of XML merger: resolve before merge
from #403 we discovered in sprint-review that the following issue was not really addressed:
* [ ] Resolve before Merging (in Load()): moving resolve into the load method

Details: The XML templates may contain variables (`$[IDE_HOME]`, `$[M2_REPO]`, etc.) that should be resolved when the workspace config is updated. Doing this after the merge has several flaws:
* If the workspace file (and not the template) itself has some variable this would also be resolved what is undesired. If legacy support for variables is enabled, even things like `${JAVA_HOME}` could be resolved accidentally if the IDE uses it inside XML config files in the workspace.
* If the variable would be used in some attribute like `<myelement id="$[JAVA_VERSION]">` in the template then on the first run of configurator the workspace is empty and this would be created and resolved. On the next run, the merger would not match `<myelement id="$[JAVA_VERSION]">` together with the resolved one from the workspace such as `<myelement id="21">` and therefore two elements would be created and after resolving the variable they would even be duplicated if the variable value did not change. This surely is an edge-case but still indicating why the current approach is somehow wrong.

As Salim told there where quite some tests breaking if this was implemented so be prepared to reserve some time for testing and analysis when implementing this story.

### LINKED ISSUE #691: XMLMerger can not handle merge of subnodes properly
### Expected behavior

As an IDEasy intellij user I expect the merger to merge my `jdktables.xml` properly.

### Actual behavior
The `XMLMerger` ignores the `name()` attribute if the parent node and the child node have the same name (not the id).

### Steps to reproduce (bug) / Use Case of feature request (enhancement)

1. Try to merge the following example:

root:
```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<application xmlns:merge="https://github.com/devonfw/IDEasy/merge" merge:strategy="COMBINE">
  <roots>
    <annotationsPath>
      <root type="composite" merge:id="name()">
        <root type="simple" name="test"/>
      </root>
    </annotationsPath>
  </roots>
</application>
```

target:
```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<application>
  <roots>
    <annotationsPath>
      <root type="composite">
        <root type="simple" name="test2"/>
      </root>
    </annotationsPath>
  </roots>
</application>
```

UPDATE/Question: What if we add ` merge:id="@name"` to the inner `root`

### Related/Dependent Issues
Blocks: #582 

### Affected version:

* IDEasy: 2024.09.002-beta-SNAPSHOT
* OS: Windows

### PR DESCRIPTION
**Pull request by hohwille**
fixes #498 and fixes #691

* fixed variable resolution bug
* resolve variables directly after loading XML (#498)
* completely refactored XmlMerger (avoid Wrappers for Element and Attr, avoid mutation of template XML)
* smart fallbacks for `merge:id`
* added caching of first `merge:id` per QName
* changed fallback for `merge:strategy` to `combine` as this is the only default that makes sense
* fixed several bugs
* solved sub-node merging issue for Intellij problems (#691)
* test enhancements
* added missing JavaDocs
* renamed all `source.xml` to `template.xml` to avoid confusion
* renamed local variables to be consistent (template vs. update vs. setup, result vs. target, workspace vs. target vs. update)

Open tasks:
- [ ] CHANGELOG needs to be added
- [ ] we could add more robustness: I had situations where I ended up with an empty XML file in my workspace. The could then not be parsed on the next run and was stuck until manual repair. An empty workspace file should be treated as if it did not exist. Also we could discuss how to deal with corrupted XML files that cannot be parsed. However, the good news is that we (or better saying @salimbouch) already have done a lot: There is robustness so a file causing an error is logged but the process is not blocked but continues. If there have been errors the user gets informed and asked to continue. Also the error logging is very precise including the absolute path of the XML files involved and the XPath within the XML document where the error happened. This is great stuff for analysing the root problem and will help us a lot in support (or prevent us from support requests as the users can help themselves).

### CODE REVIEW

File: cli/src/main/java/com/devonfw/tools/ide/environment/AbstractEnvironmentVariables.java [hunk 1]
```diff
@@ -206,7 +206,7 @@ private String resolveRecursive(String value, Object source, int recursion, Abst
     if (context.syntax == null) {
       resolved = resolveWithSyntax(value, source, recursion, resolvedVars, context, VariableSyntax.SQUARE);
       if (context.legacySupport) {
-        resolved = resolveWithSyntax(value, source, recursion, resolvedVars, context, VariableSyntax.CURLY);
+        resolved = resolveWithSyntax(resolved, source, recursion, resolvedVars, context, VariableSyntax.CURLY);
```
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/environment/AbstractEnvironmentVariables.java:209]: This was a severe bug that reverted the variables resolved in line 207

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            mergeStrategy = this; // fallback "this" will always be COMBINE
+          }
+          mergeStrategy.merge(templateChildElement, matchedResultElement, elementMatcher);
+        } else {
+          Node resultChildElement = resultElement.getOwnerDocument().importNode(templateChildElement, true);
+          resultElement.appendChild(resultChildElement);
+        }
+      } else if (XmlMergeSupport.isTextual(templateChild)) {
+        if (!templateChild.getTextContent().isBlank()) {
+          // TODO this looks wrong to me...
```
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java:111]: Needs to be discussed. Either we have to refactor this or we can clarify and just remove the comment:
```suggestion
```

File: cli/src/test/resources/xmlmerger/intellijJdkTable/template.xml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<application xmlns:merge="https://github.com/devonfw/IDEasy/merge" merge:strategy="combine">
+  <component name="ProjectJdkTable">
+    <jdk version="2" merge:id="./jdk/version[@value='OpenJDK $[JAVA_VERSION]']/.." merge:strategy="override">
```
Comment (hohwille) [inline cli/src/test/resources/xmlmerger/intellijJdkTable/template.xml:4]: for `ide-settings` I rather suggest `./jdk/version[@name='Java']/..` but for the tests I also wanted to cover and verify #498

**Comment by hohwille** [comment]
For the record: I did not change any testdata semantically. I just followed convention-over-configuration and prefer consistent usage of lowercase for `merge:strategy` enum values. We should strictly and consistently align such things both in testdata and especially in `ide-settings` (PR not yet created). Our users will take us by example. If even we do not systematically align this then we can expect a lot of mess in the projects we later have to support.

**Comment by hohwille** [comment]
CHANGELOG needs to be added

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeDocument.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NodeList;
+
+/**
+ * A wrapper for an XML {@link org.w3c.dom.Document}
+ */
+public class XmlMergeDocument {
+
+  /** The */
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeDocument.java:14]:
```suggestion
```

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    try {
+      doMerge(templateElement, resultElement, matcher);
+    } catch (XmlMergeException e) {
+      throw e;
+    } catch (RuntimeException e) {
+      throw new XmlMergeException("Merge strategy " + this + " failed on " + XmlMergeSupport.getXPath(templateElement, true), e);
+    }
+  }
+
+  protected abstract void doMerge(Element templateElement, Element resultElement, ElementMatcher matcher);
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java:73]: Please add Javadocs.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+   * @return the XPath expression.
+   */
+  static String getXPath(Element element) {
+
+    return getXPath(element, false);
+  }
+
+  /**
+   * @param element the {@link Element} for which the XPath is requested.
+   * @param includeAttributes {@code true} to also include the attributes of
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java:79]: Javadoc not complete?

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+    StringBuilder sb = new StringBuilder();
+    getXPath(sb, element, includeAttributes);
+    return sb.toString();
+  }
+
+  /**
+   * @param sb the {@link StringBuilder} where to {@link StringBuilder#append(String) append} the XML.
+   * @param element the {@link Element} with the XPath to append.
+   */
+  private static void getXPath(StringBuilder sb, Element element, boolean includeAttributes) {
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java:93]: Please adjust Javadoc, includeAttributes is missing.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+
+    removeAttriubtes(element, XmlMergeSupport::hasMergeNamespace);
+  }
+
+  /**
+   * @param element the {@link Element} from which to remove matching {@link Attr attributes}.
+   * @param filter the {@link Predicate} that {@link Predicate#test(Object) decides} if an {@link Attr attribute} shall be removed (if {@code true}) or
+   *     not.
+   */
+  static void removeAttriubtes(Element element, Predicate<Attr> filter) {
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java:213]: Typo `removeAttriubtes`, please refactor.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java [hunk 9]
```diff
@@ -151,41 +131,68 @@ public void inverseMerge(Path workspace, EnvironmentVariables variables, boolean
...
+   * @param document the XML {@link Document} to save.
+   * @param file the {@link Path} to the file where to save the XML.
+   */
   public void save(Document document, Path file) {
 
     ensureParentDirectoryExists(file);
     try {
+      //Source xsltSource = new StringSource(xslt);
+      //Templates templates = TRANSFORMER_FACTORY.newTemplates(xsltSource);
+      //Transformer transformer = templates.newTransformer();
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java:183]:
```suggestion
```

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java [hunk 10]
```diff
@@ -151,41 +131,68 @@ public void inverseMerge(Path workspace, EnvironmentVariables variables, boolean
...
       Transformer transformer = TRANSFORMER_FACTORY.newTransformer();
       transformer.setOutputProperty(OutputKeys.INDENT, "yes");
       transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2");
       transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");
-      transformer.setOutputProperty(OutputKeys.STANDALONE, "no");
+      // transformer.setOutputProperty(OutputKeys.STANDALONE, "no");
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java:188]:
```suggestion
```

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java [hunk 11]
```diff
@@ -201,7 +208,7 @@ private void removeWhitespace(Node node) {
     NodeList children = node.getChildNodes();
     for (int i = 0; i < children.getLength(); i++) {
       Node child = children.item(i);
-      if (child.getNodeType() == Node.TEXT_NODE) {
+      if (XmlMergeSupport.isTextual(child)) {
```
Comment (salimbouch) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java:211]: should CDATA also be trimmed?

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeDocument.java (hunk 4 above)
Comment (jan-vcapgemini) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeDocument.java]: Adjusted.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java (hunk 8 above)
Comment (jan-vcapgemini) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java]: Adjusted.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java (hunk 9 above)
Comment (jan-vcapgemini) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java]: Adjusted.

Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java]: I would like to remove the `merge` namespace and the indent whitespaces via a transformer stylesheet.
IMHO this would be more clean and efficient.
I will create a new issue for it.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java (hunk 11 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java:207]: I do not see any difference as all are just text.
However, we could also revert this little change and consider it as a feature so you can use CDATA to prevent trimming what makes sense to me.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java (hunk 2 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java:113]: But this is not what is implemented here.
*Every* non-empty text node will be replaced using this method and replace the text into the first non-blank text node.
So if you have
template: Text(text1) CDTATA(text2) Text(text3)
workspace: Text(text4) CDATA(text5) Text(text6)
you will create
1. iteration: Text(text1) CDTATA(text5) Text(text6)
2. iteration: Text(text2) CDTATA(text5) Text(text6)
3. iteration: Text(text3) CDTATA(text5) Text(text6)

Lets replace the TODO with a new bug issue.

Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java:113]: Created #943

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+   * @param updateChild the new text node
+   */
+  protected void replaceTextNode(Element element, Node updateChild) {
+
+    try {
+      NodeList targetChildNodes = element.getChildNodes();
+      for (int i = 0; i < targetChildNodes.getLength(); i++) {
+        Node targetChild = targetChildNodes.item(i);
+        if (XmlMergeSupport.isTextual(targetChild)) {
+          if (!targetChild.getTextContent().isBlank()) {
```
[the remark this answers is not in the document]
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java:133]: I did not implement this but it is your code from here:
https://github.com/devonfw/IDEasy/blob/6f167c101b1181be14cc8ecb199a5791ba48c46a/cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/MergeStrategy.java#L206

I will not change this in this PR and created #943 for potential improvements.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java (hunk 8 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java]: Dont know what went wrong here, but it was not fixed. I just fixed it even though this was already marked as resolved.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java (hunk 11 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMerger.java]: I changed this back to only trim TEXT and not CDATA. Thanks for pointing me to this.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    }
+  }
+
+  /**
+   * Replaces the text node in the target element with the text from the update element, otherwise appends it.
+   *
+   * @param element the element to be updated
+   * @param updateChild the new text node
+   */
+  protected void replaceTextNode(Element element, Node updateChild) {
```
[the remark this answers is not in the document]
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java]: Done.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java (hunk 7 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java]: Its a private method and the public `getXPath` method with such arguments is just above.
I just removed the JavaDoc as every developer should be able to read and understand the public method and then also get the private one called from the public method.
Having all this Copy&Paste of JavaDoc is IMHO not always making things better esp. when something changes and only parts of the redundant JavaDoc gets updated.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java (hunk 6 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeSupport.java]: Thanks for your finding - seems I was in the middle of writing this and then got distracted. I fixed this.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java (hunk 5 above)
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/XmlMergeStrategy.java:80]: Done.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/matcher/NamespaceContextFromElement.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
+package com.devonfw.tools.ide.merge.xmlmerger.matcher;
+
+import java.util.Collections;
+import java.util.Iterator;
+import javax.xml.namespace.NamespaceContext;
+
+import org.w3c.dom.Element;
+
+class NamespaceContextFromElement implements NamespaceContext {
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/matcher/NamespaceContextFromElement.java:9]: Please add Javadoc.

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/matcher/IdComputer.java [hunk 15]
```diff
@@ -44,86 +46,67 @@ public String getId() {
...
       }
     } catch (XPathExpressionException e) {
-      throw new IllegalStateException("Failed to match " + sourceElement.getXPath(), e);
+      throw new IllegalStateException("Failed to compile XPath expression " + xpath, e);
     }
   }
 
   /**
    * Builds the XPath expression for the given merge element based on the ID value.
    *
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/matcher/IdComputer.java:77]:
```suggestion
```

File: cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/matcher/IdComputer.java [hunk 16]
```diff
@@ -44,86 +46,67 @@ public String getId() {
...
-   * @param mergeElement the merge element for which to build the XPath expression
-   * @return the XPath expression as a String
+   * @param element the {@link Element} for which to build the XPath expression
+   * @return the XPath expression as a {@link String}.
    */
```
Comment (jan-vcapgemini) [inline cli/src/main/java/com/devonfw/tools/ide/merge/xmlmerger/matcher/IdComputer.java:80]:
```suggestion
```

**Review by jan-vcapgemini — APPROVED** [review]
LGTM.

**Review by salimbouch — APPROVED** [review]
LGTM

**Review by jan-vcapgemini — APPROVED** [review]
